src/rules/prefer-hooks-in-order.ts

```diff
@@ -0,0 +1,74 @@
+import { createRule, isHookCall } from './utils';
+
+const HooksOrder = [
+  'beforeAll',
+  'beforeEach',
+  'afterEach',
+  'afterAll',
+] as const;
+
+export default createRule({
+  name: __filename,
+  meta: {
+    docs: {
+      category: 'Best Practices',
+      description: 'Prefer having hooks in a consistent order',
+      recommended: false,
+    },
+    messages: {
+      reorderHooks: `\`{{ currentHook }}\` hooks should be before any \`{{ previousHook }}\` hooks`,
+    },
+    schema: [],
+    type: 'suggestion',
+  },
+  defaultOptions: [],
+  create(context) {
+    let previousHookIndex = -1;
+    let inHook = false;
+
+    return {
+      CallExpression(node) {
+        if (inHook) {
+          // Ignore everything that is passed into a hook
+          return;
+        }
+
+        if (!isHookCall(node, context.getScope())) {
+          // Reset the previousHookIndex when encountering something different from a hook
+          previousHookIndex = -1;
+        }
+
+        if (isHookCall(node, context.getScope())) {
+          inHook = true;
+          const currentHook = node.callee.name;
+          const currentHookIndex = HooksOrder.findIndex(
+            name => name === currentHook,
+          );
+
+          if (currentHookIndex < previousHookIndex) {
+            context.report({
+              messageId: 'reorderHooks',
+              node,
+              data: {
+                previousHook: HooksOrder[previousHookIndex],
+                currentHook,
+              },
+            });
+          } else {
+            previousHookIndex = currentHookIndex;
+          }
+        }
+      },
+      'CallExpression:exit'(node) {
+        if (!isHookCall(node, context.getScope()) && !inHook) {
+          // Reset the previousHookIndex when encountering something different from a hook
+          previousHookIndex = -1;
+        }
+
+        if (isHookCall(node, context.getScope())) {
```

Comment:
Again, we're doubling up our conditional check here - if you use an early return we should be able to just do the check once.